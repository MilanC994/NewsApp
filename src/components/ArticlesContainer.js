import React from 'react'
import { Dots } from 'react-activity'
import { CardColumns, Button, Container } from 'react-bootstrap'
import ArticleCard from './ArticleCard'
import useArticlesContainer from './hooks/useArticlesContainer'

function Error({ message }) {
  return (
    <div className="p-3 bg-dark rounded">
      {' '}
      <h5 className="text-danger font-weight-bold">{message}</h5>
    </div>
  )
}

function ArticlesContainer() {
  const {
    articles,
    getMoreData,
    shouldRenderLoadMoreButton,
    articlesLoading,
    articlesError,
    moreArticlesLoading,
    moreArticlesError,
  } = useArticlesContainer()
  if (articlesLoading || articlesError) {
    return (
      <div className="d-flex flex-column flex-1 align-items-center justify-content-center">
        {articlesLoading && !articlesError && <Dots color="red" size={30} />}
        {articlesError && <Error message={articlesError} />}
      </div>
    )
  }

  return (
    <>
      <CardColumns id="articles-container" className="mw-100">
        {articles
          && articles.map((article) => (
            <ArticleCard key={article.url + Math.random()} {...article} />
          ))}
      </CardColumns>
      {shouldRenderLoadMoreButton && (
        <Container className="d-flex justify-content-center align-items-center flex-column">
          {moreArticlesLoading ? <Dots color="red" size={30} />
            : (
              <Button
                variant="primary"
                id="load-more-articles-button"
                className="w-25 mb-2"
                onClick={getMoreData}
              >
                LOAD MORE
              </Button>
            )}
          {moreArticlesError && <Error message={moreArticlesError} />}
        </Container>
      )}
    </>
  )
}

export default ArticlesContainer
