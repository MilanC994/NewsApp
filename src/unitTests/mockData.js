import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import reducer from '../redux/reducer'

export const createTestStore = initialState => {
  const createStoreWithMiddleware = applyMiddleware(thunk)(createStore)
  return createStoreWithMiddleware(reducer, initialState)
}
export const mockArticles = [
  {
    author: 'Jim Halpert',
    content: 'test content 1',
    description: 'test description 1',
    publishedAt: '2021-01-02T13:25:54Z',
    source: { id: null, name: 'test src 1' },
    title: 'test title 1',
    url: 'test site url 1',
    urlToImage: 'test image url 1',
  },
  {
    author: 'Pam Beasley',
    content: 'test content 2',
    description: 'test description 2',
    publishedAt: '2021-01-02T13:25:54Z',
    source: { id: null, name: 'test src 2' },
    title: 'test title 2',
    url: 'test site url 2',
    urlToImage: 'test image url 2',
  },
  {
    author: 'Stanley Hudson',
    content: 'test content 3',
    description: 'test description 3',
    publishedAt: '2021-01-02T13:25:54Z',
    source: { id: null, name: 'test src 3' },
    title: 'test title 3',
    url: 'test site url 3',
    urlToImage: 'test image url 3',
  },
  {
    author: 'Dwight Schrute',
    content: 'test content 4',
    description: 'test description 4',
    publishedAt: '2021-01-02T13:25:54Z',
    source: { id: null, name: 'test src 4' },
    title: 'test title 4',
    url: 'test site url 4',
    urlToImage: 'test image url 4',
  },
  {
    author: 'Creed Bratton',
    content: 'test content 5',
    description: 'test description 5',
    publishedAt: '2021-01-02T13:25:54Z',
    source: { id: null, name: 'test src 5' },
    title: 'test title 5',
    url: 'testsiteurl5',
    urlToImage: 'testimageurl5',
  },
]

export const mockResponse = {
  status: 200,
  statusText: '',
  data: {
    status: 'ok',
    totalResults: 5,
    articles: mockArticles,
  },
}

export const mockState = {
  articles: [],
  searchTerm: '',
  sortBy: 'publishedAt',
  country: 'us',
  page: 1,
  totalResults: 35,
}
