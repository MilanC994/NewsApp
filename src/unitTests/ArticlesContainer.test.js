import { createTestStore, mockResponse } from './mockData'
import React from 'react'
import { mount } from 'enzyme'
import ArticlesContainer from '../components/ArticlesContainer'
import { Provider } from 'react-redux'
import { BrowserRouter as Router } from 'react-router-dom'
import { mountToJson } from 'enzyme-to-json'
import moxios from 'moxios'

beforeEach(() => {
    moxios.install()
    moxios.wait(() => {
        const request = moxios.requests.mostRecent();
        request.respondWith(mockResponse)
    })
})

afterEach(() => {
    moxios.uninstall()
})

const axiosArticlesResponse = response => {
    moxios.wait(() => {
        const request = moxios.requests.mostRecent();
        request.respondWith(response)
    })
}

describe('Articles Container Test test', () => {
    test('Renders correctly', () => {
        const initialState = {
            articles: [],
            searchTerm: '',
            sortBy: 'publishedAt',
            country:'us',
            page: 1,
            totalResults: null
        }
        const store = createTestStore(initialState) 

        const wrapper = mount(
                <Provider store={store}>
                    <ArticlesContainer />
                </Provider>
                    ).childAt(0)
        expect(mountToJson(wrapper)).toMatchSnapshot()
            })
    
    test('Renders load more button when more articles exist', () => {
        const initialState = {
            articles: [],
            searchTerm: '',
            sortBy: 'publishedAt',
            country:'us',
            page: 1,
            totalResults: 35
        }

        const store = createTestStore(initialState) 

        const wrapper = mount(
            <Provider store={store}>
                <Router>
                    <ArticlesContainer />
                </Router>
            </Provider>
        )
        const element = wrapper.find('#load-more-articles-button').at(1)   
        expect(element.exists).toBeTruthy()
       
    })
    test('Does not render load more button when more articles do not exist', () => {    
        const initialState = {
            articles: [],
            searchTerm: '',
            sortBy: 'publishedAt',
            country:'us',
            page: 1,
            totalResults: 0
        }

        const store = createTestStore(initialState) 

        const wrapper = mount(
            <Provider store={store}>
                <Router>
                    <ArticlesContainer />
                </Router>
            </Provider>
        ).at(0).at(0)
        const element = wrapper.find('#load-more-articles-button')
    
         expect(element.exists()).toBeFalsy()
       
    })

})