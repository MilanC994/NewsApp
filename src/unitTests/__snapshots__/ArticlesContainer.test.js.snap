// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Articles Container Test test Renders correctly 1`] = `
<ArticlesContainer>
  <CardColumns
    id="articles-container"
  >
    <div
      className="card-columns"
      id="articles-container"
    />
  </CardColumns>
  <hr />
</ArticlesContainer>
`;
